packages/xo-server/src/xapi/mixins/patching.js

```diff
@@ -464,4 +465,97 @@ export default {
     // host-by-host install
     throw new Error('non pool-wide install not implemented')
   },
+
+  async rollingPoolUpdate() {
+    const hosts = filter(this.objects.all, { $type: 'host' })
+    await Promise.all(hosts.map(host => host.$call('assert_can_evacuate')))
+
+    log.debug('Install patches')
+    await this.installPatches()
+
+    // Remember on which hosts the running VMs are
+    const vmsByHost = mapValues(
+      groupBy(
+        filter(this.objects.all, {
+          $type: 'VM',
+          power_state: 'Running',
+          is_control_domain: false,
+        }),
+        vm => {
+          const hostId = vm.$resident_on?.$id
+
+          if (hostId === undefined) {
+            throw new Error('Could not find host of all running VMs')
+          }
+
+          return hostId
+        }
+      ),
+      vms => vms.map(vm => vm.$id)
+    )
+
+    // Put master in first position to restart it first
+    const indexOfMaster = hosts.findIndex(host => host.$ref === this.pool.master)
+    if (indexOfMaster > 0) {
```

Comment:
How could it not be the case?

What would happen?

Reply:
If `indexOfMaster` is `0`, it just means that the master is already in first position. If it's `-1` though, it should probably throw instead of continuing, I'll change that.